### PR TITLE
Introduce Ruff Linter

### PR DESCRIPTION
**Pull request by ytausch**
Checklist
* [x] Added a ``news`` entry

This PR introduces [ruff](https://github.com/astral-sh/ruff) as a source code linter, which will ensure a consistent and modern code style.

Please see the commits to check which fixes where applied manually and which automatically.

### CODE REVIEW

File: tests/test_lint_recipe.py
Comment (h-vetinari) [inline tests/test_lint_recipe.py:1309]: If we replace this, it should be
```suggestion
        assert expected_message in lints
```

However, I see a lot of `self.assertIn` that aren't changed here, so I'm wondering what's the rationale for this change?

(if the above suggestion fails for whatever reason, match the expected message exactly)

File: conda_smithy/lint_recipe.py
Comment (ytausch) [inline conda_smithy/lint_recipe.py:501]: @h-vetinari This needed to be changed because the except clause is too broad.

Therefore I catch the only error that can happen here, `InvalidVersionSpec`, and, in one process, also print the `InvalidVersionSpec` error details because it helps debugging.

File: tests/test_lint_recipe.py
Comment (ytausch) [reply tests/test_lint_recipe.py:1309]: The lint now includes error details that are generated by our dependency `conda` (see the other comment in which I tagged you). Therefore, your suggestion fails. And also, I did not want to match the expected message exactly because `conda` could change their error message slightly and then our tests fails.

**Review by beckermr — CHANGES_REQUESTED** [review]
I'd be in favor of expanding the ruleset for ruff in the pyproject.toml

```
[tool.ruff.lint]
select = ["E", "F", "W"]
preview = true
```

Thoughts everyone?

**Comment by ytausch** [comment]
@beckermr What do you mean by "expanding the ruleset"?

My PR already contains more rulesets than the default (and more than you propose):

https://github.com/conda-forge/conda-smithy/blob/12f4a1fa107b75069d5380e166b2d5d41cbb6ed9/pyproject.toml#L36-L52

Also, what is the rationale behind your proposal to enable `preview` mode? Preview mode contains [unstable features](https://docs.astral.sh/ruff/preview/), which is not something we want here IMO.

File: conda_smithy/anaconda_token_rotation.py
Comment (jakirkham) [inline conda_smithy/anaconda_token_rotation.py:366]: Out of curiosity, do we know what led to this change?

File: conda_smithy/lint_recipe.py
Comment (jakirkham) [inline conda_smithy/lint_recipe.py:31]: Why is it splitting these into 2 `import`s from the same module? This seems overly verbose. Is there an option to fix this?

File: conda_smithy/lint_recipe.py
Comment (jakirkham) [inline conda_smithy/lint_recipe.py:494]: Think we can simplify this further

```suggestion
    version = str(package_section.get("version", ""))
    if version:
```

File: conda_smithy/lint_recipe.py
Comment (jakirkham) [inline conda_smithy/lint_recipe.py:496]:
```suggestion
            VersionOrder(version)
```

File: tests/conftest.py
Comment (jakirkham) [inline tests/conftest.py:9]: Think the idea here was to group `import`s and then `from...import...`s and separate those out

Does `ruff` have a way to do this?

File: tests/test_condaforge_config_schema.py
Comment (jakirkham) [inline tests/test_condaforge_config_schema.py:4]: Interesting here it groups some `from...import...`s with `import`s and splits some out. Or is it just trying to group external `import`s separately from internal ones?

**Review by jakirkham — COMMENTED** [review]
Thanks Yannik! 🙏

Generally this looks good. Had a few questions/observations below

File: conda_smithy/lint_recipe.py
Comment (jaimergp) [reply conda_smithy/lint_recipe.py:31]: I think that's part of the black/isort standard too. `as` imports have their own line 🤷

File: .pre-commit-config.yaml
Comment (jaimergp) [inline .pre-commit-config.yaml:12]: We should remove black if we have ruff now.

File: tests/conftest.py
Comment (jaimergp) [reply tests/conftest.py:9]: From what I've seen, black/isort distinguish three blocks (one blank line of separation):

- stdlib
- 3rd party
- the package itself

Then within each block, we'll see the `import` lines first, followed by `from X import Y`, alphabetically. No blank line.

**Comment by jaimergp** [comment]
Awesome! I suggest that when we merge this one, we do it via Squash, so then we can easily omit this PR in the blame view via that github file whose name I can't remember :)

File: conda_smithy/anaconda_token_rotation.py
Comment (ytausch) [reply conda_smithy/anaconda_token_rotation.py:366]: The ruff linter `I` rules sort imports with `isort`. Does that answer your question?

File: conda_smithy/lint_recipe.py
Comment (ytausch) [reply conda_smithy/lint_recipe.py:31]: Yes.

File: conda_smithy/lint_recipe.py
Comment (ytausch) [reply conda_smithy/lint_recipe.py:494]: I propose to stay defensive here and crash intentionally if `version` is an empty string (as opposed to `None`). Open to split the `:=` assignment though if you think this is too fancy.

File: conda_smithy/lint_recipe.py
Comment (ytausch) [reply conda_smithy/lint_recipe.py:496]: see above, I think this relates to the change above

File: tests/conftest.py
Comment (ytausch) [reply tests/conftest.py:9]: I also think so. The reasoning behind `isort` makes sense and we don't need to change something about it IMO.

File: tests/test_condaforge_config_schema.py
Comment (ytausch) [reply tests/test_condaforge_config_schema.py:4]: This is exactly what happens.

File: .pre-commit-config.yaml
Comment (ytausch) [reply .pre-commit-config.yaml:12]: Actually, I think the ruff linter is orthogonal to `black` as a formatter. There is also the ruff formatter (which is a Black alternative) but it is currently disabled. Using the ruff linter together with Black is a [common use case](https://docs.astral.sh/ruff/faq/#is-the-ruff-linter-compatible-with-black).

Open to switch to the ruff formatter in a separate PR if you think that makes sense - but for now, I propose to stick with Black.

Comment (beckermr) [reply .pre-commit-config.yaml:12]: agreed on using both ruff and black

**Comment by jaimergp** [comment]
Any thoughts about the rationale for the preview mode exposed in https://github.com/conda-forge/conda-forge-webservices/pull/616?

**Comment by beckermr** [comment]
I like all of the flake8 rules so I like preview mode, but that is me.

File: .pre-commit-config.yaml
Comment (beckermr) [reply .pre-commit-config.yaml:12]: Actually I am confused now. Do we still get autofixes from ruff if we don't use it as a formatter?

Comment (ytausch) [reply .pre-commit-config.yaml:12]: Yes, we do.

**Comment by ytausch** [comment]
I would be fine with enabling preview mode if you think that's necessary - is there anything else you want to debate about this PR? All open discussions are solved from my side I want this to get merged.

**Comment by ytausch** [comment]
@conda-forge/core

File: conda_smithy/lint_recipe.py
Comment (h-vetinari) [reply conda_smithy/lint_recipe.py:495]: I like John's suggestion much better.

> I propose to stay defensive here and crash intentionally if `version` is an empty string (as opposed to `None`).

That is way too subtle to go without a comment, but in any case, I don't think we need to: We should make the version check unconditional (even for `""`), which should be the best of both worlds AFAICT:
```
version = str(package_section.get("version", ""))
try:
    VersionOrder(version)
except InvalidVersionSpec as e:
    ...
```

I'm not aware of any feedstocks without a version; unless I'm overlooking a use-case here, I'd like to just get rid of that corner case completely.

**Review by h-vetinari — COMMENTED** [review]
Basically LGTM, but some small things left.

File: conda_smithy/lint_recipe.py
Comment (ytausch) [reply conda_smithy/lint_recipe.py]: How about this solution? It takes your approach but makes it super-obvious what is going on.

0c04145

Comment (h-vetinari) [reply conda_smithy/lint_recipe.py]: Fine by me. There are two tests that now produce a new linter hint that needs to be handled.

Comment (ytausch) [reply conda_smithy/lint_recipe.py]: Done.

**Comment by ytausch** [comment]
@h-vetinari @beckermr @jaimergp @jakirkham Can we please get this through?

**Review by beckermr — APPROVED** [review]

**Comment by beckermr** [comment]
Let's wait for a few more reviews and then merge.

File: conda_smithy/github.py
Comment (jaimergp) [inline conda_smithy/github.py:12]: We are not being too consistent with relative / absolute imports, I think

File: news/1919-ruff.rst
Comment (jaimergp) [inline news/1919-ruff.rst:7]:
```suggestion
* Introduced ``ruff`` as pre-commit linter. (#1919) 
```

**Review by jaimergp — APPROVED** [review]
Just some comments that came to mind while reading the changes. The only blocking one is the PR number in the news file.

File: conda_smithy/anaconda_token_rotation.py
Comment (ytausch) [reply conda_smithy/anaconda_token_rotation.py:142]: I would like, however `rotate_anaconda_token` is in such a bad condition that I don't want to touch anything which is not 1) strictly ruff-lrelated 2) a complete refactoring of the method, or removal (I could not figure out where the CLI command is used)

2) is out of scope for this PR, so I created a separate issue #1962.

File: conda_smithy/configure_feedstock.py
Comment (ytausch) [reply conda_smithy/configure_feedstock.py:21]: #1963

File: conda_smithy/configure_feedstock.py
Comment (ytausch) [reply conda_smithy/configure_feedstock.py:325]: In general, yes. And there is even a ruff rule to check that (which is currently disabled): [G004](https://docs.astral.sh/ruff/rules/logging-f-string/)

I personally prefer f-strings for logging because I think they are easier to read, and usually there should not be much performance to gain because calculating the string representations of the objects should be fast anyway.

I found https://stackoverflow.com/a/54368109, so it seems like I am in a minority, so if you want to, I am totally open to adding the G-rules enforcing logging with deferred positional arguments. Let me know what you think.

File: conda_smithy/feedstock_tokens.py
Comment (ytausch) [reply conda_smithy/feedstock_tokens.py]: done in 0840cc7

File: conda_smithy/github.py
Comment (ytausch) [reply conda_smithy/github.py:12]: This is how the new file looks like:

https://github.com/conda-forge/conda-smithy/blob/d8a8d8dab3baa97a61374bc14764995931a30053/conda_smithy/github.py#L1-L14

Seems pretty ordered to me. Note https://github.com/conda-forge/conda-smithy/pull/1919#discussion_r1593447856

Comment (jaimergp) [reply conda_smithy/github.py:12]: Sometimes we are doing `from .utils import X` and sometimes `from conda_smithy.utils import X`.

Compare this line to this [other one](https://github.com/conda-forge/conda-smithy/pull/1919/files#diff-27f5c1fcf8e15b2b01490459c0c6589f02d699d19110b6a9a80d66b034f6fc8bR19)

File: conda_smithy/configure_feedstock.py
Comment (jaimergp) [reply conda_smithy/configure_feedstock.py:325]: No strong feelings here. My style is usually deferring as long as possible in logging out of habit so this reads weird, but I'm not going to oppose. Whatever we use, we should be consistent though. So pick one and enforce a rule, it's what I'd say.

Comment (ytausch) [reply conda_smithy/configure_feedstock.py:331]: Alright, let's enforce the G rules then.

Comment (ytausch) [reply conda_smithy/configure_feedstock.py]: 5dd60c1ccaa8eaa6e0c4a1d24972b50541dff88f

File: conda_smithy/github.py
Comment (ytausch) [reply conda_smithy/github.py:12]: Good point. My proposal: Let's ban all relative imports with the `TID` ruleset and `flake8-tidy-imports.ban-relative-imports = "all"`.

I've done this in 4902d5d.

Quote from [PEP 8](https://peps.python.org/pep-0008/#imports):

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on sys.path):
```python
import mypkg.sibling
from mypkg import sibling
from mypkg.sibling import example
```
> However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:
```python
from . import sibling
from .sibling import example
```

So I think we're in line with PEP 8 since the package layout is not complex at all.

**Comment by ytausch** [comment]
@jaimergp

**Comment by ytausch** [comment]
Rebasing this over and over again is quite tedious so I'll wait until we have found a consensus about the rules we want to enforce and then I can do this one last time. Happy if we finally close this after almost 3 months.

**Comment by beckermr** [comment]
We have two approvals from core , so I'd be happy to merge today.

**Comment by ytausch** [comment]
To make rebasing easier and the commit history cleaner, I reran ruff with the final configuration, and reapplied everything we have discussed here. From my side, this should be fine now.

**Comment by beckermr** [comment]
We have two test failures. After those are fixed, we can merge.

**Comment by ytausch** [comment]
Done.

**Comment by wolfv** [comment]
Awesome job, @ytausch

**Pull request merged**
